### PR TITLE
Add variable CLI command design

### PR DESCRIPTION
**Pull request by sigalsax**
### What does this PR do?

Adds design for variable CLI command

### CODE REVIEW

File: design/variable_command_design.md [hunk 1]
```diff
@@ -0,0 +1,181 @@
...
+
+2. Should we return error codes? 404 if not found on secrets?
+
+### Delivery plan
+
+1. Ensure UTs and integration tests exist and are GA quality (3 days)
+   1. Sit with Elad to map out current tests and what is missing
+   2. Implement missing tests (if required)
+2. Improve error handling experience (3 days)
+   1. ***Blocker*** - need to get full UX outlines from Sharon + Inbal
```
Comment (InbalZilberman) [inline design/variable_command_design.md:177]: I think we agreed on \" and spaces
conjur variable get "var1" "var2"
OK?
Ill update the Epic also

File: design/variable_command_design.md [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+### Open questions
+
+1. If extra arguments are provided, what should be the experience? See second to last flow.
+
+2. Should we return error codes? 404 if not found on secrets?
+
+### Delivery plan
+
+1. Ensure UTs and integration tests exist and are GA quality (3 days)
+   1. Sit with Elad to map out current tests and what is missing
```
Comment (InbalZilberman) [inline design/variable_command_design.md:174]: I think it is best to write QA & PO  :)

File: design/variable_command_design.md [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+$ conjur-cli variable set foo/bar mysecret sfdfs
+
+## Current
+usage: conjur-cli [-h] [-v] [-l URL] [-a ACCOUNT] [-c CA_BUNDLE] [--insecure] [-u LOGIN_ID] [-k API_KEY] [-p PASSWORD] [-d]
+         [--verbose]
+         {list,variable,policy} ...
+conjur-cli: error: unrecognized arguments: sfdfs
+
+## Should be:
+Should the last input be ignored? Should it take only the arguments it needs (first 2) as we do in Ruby CLI?
```
Comment (InbalZilberman) [inline design/variable_command_design.md:148]: yes. Ignored

File: design/variable_command_design.md [hunk 4]
```diff
@@ -0,0 +1,181 @@
...
+mysecret%
+```
+
+*Get a secret without inputting which secret to get*
+
+```bash
+$ conjur-cli variable get __
+
+## Current 
+usage: conjur-cli variable get [-h] variable_id [variable_id ...]
```
Comment (InbalZilberman) [inline design/variable_command_design.md:98]: usage: conjur-cli variable get [-h] **"**variable_id**"** [**"**variable_id**"** ...]

File: design/variable_command_design.md [hunk 5]
```diff
@@ -0,0 +1,181 @@
...
+
+### Delivery plan
+
+1. Ensure UTs and integration tests exist and are GA quality (3 days)
+   1. Sit with Elad to map out current tests and what is missing
+   2. Implement missing tests (if required)
+2. Improve error handling experience (3 days)
+   1. ***Blocker*** - need to get full UX outlines from Sharon + Inbal
+3. Write and get approval on UX logs (2 days)
+4. Write proper documentation in README.md of the repository (2 days)
```
Comment (InbalZilberman) [inline design/variable_command_design.md:179]: 5. Write proper documentation in confluence and open doc ticket

File: design/variable_command_design.md [hunk 6]
```diff
@@ -0,0 +1,181 @@
...
+*Get a secret without inputting which secret to get*
+
+```bash
+$ conjur-cli variable get __
+
+## Current 
+usage: conjur-cli variable get [-h] variable_id [variable_id ...]
+
+## Should be
+Error: Command 'variable' requires the following arguments: variable_id
```
Comment (sgnn7) [inline design/variable_command_design.md:101]: This command accepts one-or-more variable IDs so the description needs to describe that wording. I do think that the original version of the help here is much more helpful for a CLI user.

**Review by sgnn7 — COMMENTED** [review]
@sigalsax Left a few comments

File: design/variable_command_design.md (hunk 1 above)
Comment (sigalsax) [reply design/variable_command_design.md:177]: yes but I am talking about the actual help screen, info, and error statements here

File: design/variable_command_design.md (hunk 4 above)
Comment (sigalsax) [reply design/variable_command_design.md:98]: this the current flow and I am suggesting below:
`Error: Command 'variable' requires the following arguments: variable_id`

File: design/variable_command_design.md [hunk 7]
```diff
@@ -0,0 +1,181 @@
+# Dockerless CLI - variable
+
+This document is the design document for the variable CLI command for the Dockerless project. The functionality for the variable CLI command already exists so all that is left is the following:
```
[the remark this answers is not in the document]
Comment (sigalsax) [reply design/variable_command_design.md:3]: The line or the sentence?

File: design/variable_command_design.md [hunk 8]
```diff
@@ -0,0 +1,181 @@
...
+*Set secrets*
+
+```bash
+$ conjur-cli variable set foo/bar mysecret
+
+## Current
+Value set: 'foo/bar'
+
+## Should be
+Success! Data written to: 'foo/bar'
```
[the remark this answers is not in the document]
Comment (sigalsax) [reply design/variable_command_design.md:82]: Out of curiousity, what makes this non-machine readable? The `''` and the `!`?

File: design/variable_command_design.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (sgnn7) [reply design/variable_command_design.md:3]: Line itself. This is easily over 100 characters so it needs breaking down into multiple lines..

File: design/variable_command_design.md (hunk 8 above)
[the remark this answers is not in the document]
Comment (sgnn7) [reply design/variable_command_design.md:82]: If you try to give me a way to parse the variable name from each of those outputs, you will see that the suggested use case requires careful counting vs otherwise. Both can be extracted the same way but the original output is less prone to error during script development.

File: design/variable_command_design.md [hunk 9]
```diff
@@ -0,0 +1,181 @@
...
+
+```bash
+$ conjur-cli variable boo
+
+## Current
+usage: conjur-cli variable [-h] {get,set} ...
+conjur-cli variable: error: argument action: invalid choice: 'boo' (choose from 'get', 'set')
+
+## Should be
+Error: Command 'variable' requires a subcommand get, set
```
[the remark this answers is not in the document]
Comment (sigalsax) [reply design/variable_command_design.md:54]: So it seems that argparse is a wrapper for our error handling and help screens giving us the form
```
usage: myprog.py [-h] xyz

positional arguments:
  xyz   ...

optional arguments:
  -h, --help  ..
```
Requiring us to switch this out for something else, or build one ourself (please no), or create a wrapper for catching these methods and returning them in the format suggested above?

[the remark this answers is not in the document]
Comment (sigalsax) [reply design/variable_command_design.md:54]: After some research I found [this](https://stackoverflow.com/questions/35847084/customize-argparse-help-message) that could help us get around the format.

TL;DR, we could do the following:
1. Suppress the help given to us and providing our own
2. Override optional arguments/positional (_optional_argument, _position_argument) but these are internal attributes and may lead to breaking changes if the module is updated in this area
3. Provide a wrapper for the usage class as explained in the link

These are some idea but requires more research. What do you think?

File: design/variable_command_design.md (hunk 8 above)
[the remark this answers is not in the document]
Comment (sigalsax) [reply design/variable_command_design.md:82]: why careful counting if we split at the spaces and take the last index. Can you explain a bit more about what you mean by this or why current output makes it less error prone?

[the remark this answers is not in the document]
Comment (sigalsax) [reply design/variable_command_design.md:78]: As we spoke about that the writing variables isn’t critical as is reading of variable be machine-readable so it’s ok

File: design/variable_command_design.md [hunk 10]
```diff
@@ -0,0 +1,179 @@
...
+*Set secrets*
+
+```bash
+$ conjur-cli variable set foo/bar mysecret
+
+## Current
+Value set: 'foo/bar'
+
+## Should be
+Success! Data written to: 'foo/bar'
```
Comment (sigalsax) [inline design/variable_command_design.md:78]: @sgnn7 what about `Success! Data set to: ''`?

File: design/variable_command_design.md (hunk 6 above)
Comment (sigalsax) [reply design/variable_command_design.md:97]: This was a mistake, the current error is:
```
usage: conjur-cli variable get [-h] variable_id [variable_id ...]
conjur-cli variable get: error: the following arguments are required: variable_id
```

File: design/variable_command_design.md (hunk 10 above)
Comment (sgnn7) [reply design/variable_command_design.md:78]: Sounds good to me

File: design/variable_command_design.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (sigalsax) [reply design/variable_command_design.md:3]: I use [Typora](https://typora.io/) when I write MD so I can't measure how it looks in raw markdown. Is this critical because MDs are rendered in GH and won't be an excessive long line as we see when in raw

**Review by InbalZilberman — APPROVED** [review]
Great work! thank you

**Review by eladkug — APPROVED** [review]
LGTM
